### DISCUSSION TITLE
[Script request] Joplin LXC

### DISCUSSION THREAD

**Discussion by FaboulousSan** [community-scripts/ProxmoxVE#444] [Request script]
### Application Name

Joplin

### Website

https://github.com/laurent22/joplin

### Description

Hello, I've been looking for quite some time for a script that allows using/installing Joplin Server without Docker. I found some documentation [here](https://syn-flood.com/posts/2022/12/30/joplin-lxc-2.9.html), but so far, I haven't made much progress. Perhaps the community could once again do an excellent job. If I make progress on my end, I'll keep you posted.

### Due Diligence

- [X] I have searched existing [scripts](https://community-scripts.github.io/Proxmox/scripts) and found no duplicates.
- [X] I have searched existing [discussions](https://github.com/community-scripts/ProxmoxVE/discussions?discussions_q=) and found no duplicate requests.
- [X] This is not a game-related request.

**Comment by crivchri** [comment]
I would love to have a script for Joplin!
I have tried without success to install Joplin in a LXC, but I cannot get it to work.
Would be happy to help!

**Comment by michelroegl-brunner** [comment]
I dont really know if this should be added. This seems like a really hacky way to do this and is prone to breaking and would be a pain to maintain and update. I also dont see anythin in the official docs regarding a server intall.

**Reply by tremor021** [reply]
i took a look at it. If its the server that this request is meant for, its just `joplin server start` command

**Reply by crivchri** [reply]
Hi @michelroegl-brunner, hi @tremor021,

Thank you for your answers.

The 'official' way to have the Joplin Server running is to use Docker (https://github.com/laurent22/joplin/blob/dev/docker-compose.server.yml) and the image from https://hub.docker.com/r/joplin/server. But I am not willing to use Docker on Proxmox and would prefer go with a LXC for the Joplin Server.

I would be very happy to get your help for this!

**Reply by crivchri** [reply]
Could someone please advise me on how to create an install script for Joplin Server on a simple Linux LXC (without Docker)?

I am not experienced enough to understand how get all necessary commands out of the officiel Docker image and reproduce the same process without Docker in my LXC.

Thanks a lot for any help!

**Comment by tremor021** [comment]
Sorry for waiting this long @FaboulousSan @crivchri. The script is up on dev repo, not finished yet but it can be installed.
Depending on your host speed, the build process can take some time. If you can test the script and see if it installs fine, you can do it here:

`bash -c "$(curl -fsSL https://raw.githubusercontent.com/community-scripts/ProxmoxVED/main/ct/joplin-server.sh)"`

**Reply by crivchri** [reply]
Hi @tremor021,

Amazing, I have tried your script and it works perfectly! Thank you very much for this.

Can I ask how you did to make this works (I have tried many scripts for weeks with no success)?
Would it also work with PostgreSQL 17?
Will this script be added to the repository on https://helper-scripts.com ?

Kind regards

**Reply by tremor021** [reply]
I mean, it't not really hard to make it work,, I just didn't had time to get to it earlier. I just translated the server Dockerfile to a bash script.
I've tested it only with PSQL 16, but it should work with PSQL 17, I just have to test it first.
Yes, after we finish testing, it will be included in the official repository.

**Reply by crivchri** [reply]
Hi @tremor021,

Can I ask how you proceed to translate a Dockerfile to a bash script?

Is this somehow automated?

Thank you and for your help.
Kind regards

**Reply by tremor021** [reply]
Its just knowing what Dockerfile commands are and making them into bash commands.
Once you know what Dockerfiles are and how they work, the rest is not a problem

**Reply by crivchri** [reply]
Hi @tremor021,

Thank you for your explanation.
Do you know of a tutorial that explains this Dockerfile > Bash conversion and the correspondence between Dockerfile commands and Bash commands?

Kind regards

**Reply by tremor021** [reply]
https://docs.docker.com/get-started/docker-concepts/building-images/writing-a-dockerfile/

THey have a great video on dockerfiles. Bash on the other hand you have to know kind of

**Reply by crivchri** [reply]
Hi @tremor021,

Thank for the info, I will have a look at it. 😃 

I have noticed that the LXC is created with a 20 GB disk and in my case 50% is already in use, where my 'joplin' database is only about 2 GB. What is taking so much storage space in the LXC?

Regarding the install script, would it be possible to deploy the LXC with Debian 13 'Trixie'?

Kind regards

**Reply by tremor021** [reply]
Thats NodeJS for you right there. The application is built from source, so NodeJS pulls and compiles bunch of dependencies that cannot be removed safely. I tried various methods to slim down the app but they all make application to not run.

**Reply by Pacers31Colts18** [reply]
@tremor021 

What's the default login creds? I go to the web UI and am greeted with email and password, but I've not set that up anywhere.

**Reply by MickLesk** [reply]
Read Dev Website

https://community-scripts.github.io/ProxmoxVED/scripts?id=joplin-server

**Reply by crivchri** [reply]
> Thats NodeJS for you right there. The application is built from source, so NodeJS pulls and compiles bunch of dependencies that cannot be removed safely. I tried various methods to slim down the app but they all make application to not run.

Hi @tremor021,

On this topic, I had a deeper look at the [Dockerfile.server](https://github.com/laurent22/joplin/blob/dev/Dockerfile.server) file and I noticed that it contains two parts, which seems to do the trick and keeps the container small.

The first part is creating a `builder` container, which is only aimed to build from the source and this container is then not kept.

The second part is creating the final container, where only the content of the `/build/packages` folder is copied from the `builder` container (i.e. `COPY --chown=$user:$user --from=builder /build/packages /home/$user/packages`).

Do you have an idea how this behavior could be reproduced with the LXC script?

**Reply by crivchri** [reply]
Hi @tremor021,

I adapted your install script in a dirty way, inspired from the Dockerfile (as mentioned in my previous post), and I managed to reduce the size of the container by about 4.4 GB (7.2 GB instead of 11.6 GB), which is 38% less.

I adapted your install script in a dirty way, taking inspiration from the second part of the Dockerfile (as mentioned in my previous message), and I managed to reduce the size of the container by about 4.4 GB (7.2 GB instead of 11.6 GB), which is a 38% less size.

For a container created with your script, I created a `tar.gz` archive of the `/opt/joplin-server/packages` folder. Then I copied the `tar.gz` archive to my NAS. Then in the script updated from yours (see below), I downloaded this archive with `curl` and extracted it with `tar` inside a new container. The rest of your script was not changed, except commenting out the build part.

Do you have an idea how we could combine your script with my concept to have at the end one smaller container without the build files)?

Kind regards.

```shell
#!/usr/bin/env bash

# Copyright (c) 2021-2025 community-scripts ORG
# Author: Slaviša Arežina (tremor021)
# License: MIT | https://github.com/community-scripts/ProxmoxVE/raw/main/LICENSE
# Source: https://joplinapp.org/

source /dev/stdin <<<"$FUNCTIONS_FILE_PATH"
color
verb_ip6
catch_errors
setting_up_container
network_check
update_os

msg_info "Installing Dependencies"
$STD apt-get install -y \
    git \
    rsync
msg_ok "Installed Dependencies"

PG_VERSION="17" setup_postgresql
NODE_VERSION=22 NODE_MODULE="yarn@latest,npm@latest,pm2@latest" setup_nodejs
mkdir -p /opt/pm2
export PM2_HOME=/opt/pm2
$STD pm2 install pm2-logrotate
$STD pm2 set pm2-logrotate:max_size 100MB
$STD pm2 set pm2-logrotate:retain 5
$STD pm2 set pm2-logrotate:compress tr

msg_info "Setting up PostgreSQL Database"
DB_NAME=joplin
DB_USER=joplin
DB_PASS="$(openssl rand -base64 18 | cut -c1-13)"
$STD sudo -u postgres psql -c "CREATE ROLE $DB_USER WITH LOGIN PASSWORD '$DB_PASS';"
$STD sudo -u postgres psql -c "CREATE DATABASE $DB_NAME WITH OWNER $DB_USER ENCODING 'UTF8' TEMPLATE template0;"
$STD sudo -u postgres psql -c "ALTER ROLE $DB_USER SET client_encoding TO 'utf8';"
$STD sudo -u postgres psql -c "ALTER ROLE $DB_USER SET default_transaction_isolation TO 'read committed';"
$STD sudo -u postgres psql -c "ALTER ROLE $DB_USER SET timezone TO 'UTC'"
{
    echo "Joplin-Credentials"
    echo "Joplin Database User: $DB_USER"
    echo "Joplin Database Password: $DB_PASS"
    echo "Joplin Database Name: $DB_NAME"
} >>~/joplin.creds
msg_ok "Set up PostgreSQL Database"

# REMOVED
# fetch_and_deploy_gh_release "joplin-server" "laurent22/joplin" "tarball" "latest"

msg_info "Setting up Joplin Server (Patience)"
LOCAL_IP=$(hostname -I | awk '{print $1}')

# ADD: NOT SURE IF NEEDED
mkdir -p /opt/joplin-server

cd /opt/joplin-server

# REMOVED
# sed -i "/onenote-converter/d" packages/lib/package.json
# $STD yarn config set --home enableTelemetry 0
# export BUILD_SEQUENCIAL=1
# $STD yarn install --inline-builds

# COPY FROM NAS (DIRTY TEST)
curl -fsSL https://<MY NAS IP>/scripts/packages.tar.gz -o /opt/joplin-server/packages.tar.gz
tar -xvzf packages.tar.gz --no-same-owner

cat <<EOF >/opt/joplin-server/.env
PM2_HOME=/opt/pm2
NODE_ENV=production
APP_BASE_URL=http://$LOCAL_IP:22300
APP_PORT=22300
DB_CLIENT=pg
POSTGRES_PASSWORD=$DB_PASS
POSTGRES_DATABASE=$DB_NAME
POSTGRES_USER=$DB_USER
POSTGRES_PORT=5432
POSTGRES_HOST=localhost
EOF
msg_ok "Setup Joplin Server"

msg_info "Setting up Service"
cat <<EOF >/etc/systemd/system/joplin-server.service
[Unit]
Description=Joplin Server Service
After=network.target

[Service]
Type=simple
WorkingDirectory=/opt/joplin-server/packages/server
EnvironmentFile=/opt/joplin-server/.env
ExecStart=/usr/bin/yarn start-prod
Restart=on-failure
RestartSec=5

[Install]
WantedBy=multi-user.target
EOF
systemctl enable -q --now joplin-server
msg_ok "Service Setup"

motd_ssh
customize

msg_info "Cleaning up"
$STD apt-get -y autoremove
$STD apt-get -y autoclean
msg_ok "Cleaned"
```
